gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/service/user/impl/UserServiceImpl.java

```diff
@@ -188,6 +194,11 @@ public Single<RegistrationResponse> register(Client client, User user, io.gravit
                         .switchIfEmpty(Maybe.error(new UserProviderNotFoundException(source)))
                         .flatMapSingle(userProvider -> userProvider.create(convert(user)))
                         .flatMap(idpUser -> {
+                            //noinspection ResultOfMethodCallIgnored
+                            passwordHistoryService
+                                    .addPasswordToHistory(DOMAIN, domain.getReferenceId(), idpUser.getId(), idpUser.getCredentials(), principal, getPasswordSettings(client))
```

Comment:
Do we have to use the `idpUser.getId()` or the `user.getId()` ?
I was thinking of the `user.getId()` but maybe you have good arguments in favor of the `idpUser.getId()`.

Reply:
I could be wrong but I was working on the assumption that the idp assigns the ID to the new user, so we have to wait until the user is created to get that assigned ID.

Reply:
IMHO, the only safe ID we should use as foreign key is the one of the `Users` collection (`user.getId()`).
The idpUser.getId() is used to initialize the externalId attribute of the user but as we do not control all the info provided by the IDPs and specially the UserMapper, we can't guaranty that the value will remains the same along the user lifecycle.

That's why I think the user ID should be used instead of the idpUser ID.